ENH: Add gradient-based optimization for fitting GLM's

kshedden opened the pull request:
closes #1736 and #1958, see also #1788 

This is a PR to allow GLM's to be fit using scipy optimizers.  It also fills some other gaps and corrects a few inconsistencies in the GLM implementation.  Here are the main points:

* The method keyword argument to fit can be IRLS (as before), or the name of any scipy optimizer.

* Several deriv2 implementations that were missing in family.py were falling back to numerical differentiation, which leads to issues with run time and big inaccuracies in some cases.  I have added most of the missing analytic deriv2 functions.

* I added tests for most of the links and associated functions (inverses, derivatives, etc.).

* I resolved inconsistencies in some of the signatures for the negative binomial family.

* I added a likelihood function that takes params as an argument, renaming the likelihood that takes the fitted means as loglike_mu.

* I added tests comparing IRLS fits to Newton fits for 6 families.  The tests all run clean for me.

josef-pkt commented:
@kshedden   Did you do a rebase of this PR?
Looks good. From a quick read through the changes, I didn't see any problems and it looks ready to merge.

Two things I'd like to look at, are how starting values are determined and whether we can copy a cov_type sandwich test to use the new fit option/method.

kshedden commented:
Yes,  I rebased it.  A few things in the tests had developed conflicts with the current master.

I also just changed the sample size in the test function, for speed and to perhaps catch any problems. 

I think it's good to go.  After merge we can take a look at starting values and cov_type calculations.  My main use for this now is to be able to implement fit_regularized for GLM.

josef-pkt commented on statsmodels/genmod/families/family.py:
```diff
@@ -1275,14 +1279,9 @@ def loglike(self, endog, lin_pred=None):
+        lin_pred = self._link(mu)
+        constant = special.gammaln(endog + 1/self.alpha) - special.gammaln(endog+1)\
+                    -special.gammaln(1/self.alpha)
         exp_lin_pred = np.exp(lin_pred)
```
Is there a bug .here?  Why do we take exp of lin_pred instead of `link.inverse(linpred)` i.e. mu
looks to me the loglike assumes log link and doesn't depend on actual link

I haven't verified, this just looks suspicious to me while reading. And it's not in this PR

josef-pkt commented:
I'm merging this now. 

follow up to check: negative binomial loglike looks suspicious to me

josef-pkt commented:
@kshedden   If you have a negative binomial glm example readily available then you could check the loglike by comparing the fits between IRLS and optimize, the parameter estimates should be same for log link but different for other links (I guess power link would work), if my suspicion is correct.
